### PR TITLE
Added ability to addMedia for non existing model

### PR DESCRIPTION
**Pull request by ModernArtisan**
Now you can attach medias for non existing model records.
Can provide some improvements, what do you think?

### CODE REVIEW

File: src/FileAdder/FileAdder.php [hunk 1]
```diff
@@ -331,4 +324,30 @@ protected function sanitizeFileName(string $fileName) : string
...
+            $class = get_class($this->subject);
+
+            $class::created(function ($model) {
+                if (! empty($model->unsavedMedias)) {
+                    foreach ($model->unsavedMedias as $unsavedMedia) {
+                        $model->media()->save($unsavedMedia);
+                        $this->filesystem->add($this->pathToFile, $unsavedMedia, $this->fileName);
+                    }
+                }
+            });
```
Comment (freekmurze) [inline src/FileAdder/FileAdder.php]: Add an early return here, so we can get rid of the `else` statement.

File: src/FileAdder/FileAdder.php [hunk 2]
```diff
@@ -331,4 +324,30 @@ protected function sanitizeFileName(string $fileName) : string
     {
         return str_replace(['#', '/', '\\'], '-', $fileName);
     }
+
+    /**
+     * Attach media to the model.
+     *
+     * @param $media
+     */
+    protected function attachMedia($media)
```
Comment (freekmurze) [inline src/FileAdder/FileAdder.php]: Add a `Media` type hint here

File: src/FileAdder/FileAdder.php [hunk 3]
```diff
@@ -331,4 +324,30 @@ protected function sanitizeFileName(string $fileName) : string
     {
         return str_replace(['#', '/', '\\'], '-', $fileName);
     }
+
+    /**
+     * Attach media to the model.
```
Comment (freekmurze) [inline src/FileAdder/FileAdder.php]: Let's lose the docblock: we can typehint the parameter and the function name is clear enough.

File: src/HasMedia/HasMediaTrait.php [hunk 4]
```diff
@@ -23,6 +23,9 @@
     /** @var bool */
     protected $deletePreservingMedia = false;
 
+    /** @var array */
+    public $unsavedMedias = [];
```
Comment (freekmurze) [inline src/HasMedia/HasMediaTrait.php]: rename this to `unsavedMediaItems`

File: src/FileAdder/FileAdder.php [hunk 5]
```diff
@@ -331,4 +324,30 @@ protected function sanitizeFileName(string $fileName) : string
...
+     */
+    protected function attachMedia($media)
+    {
+        if (! $this->subject->exists) {
+            $this->subject->unsavedMedias[] = $media;
+
+            $class = get_class($this->subject);
+
+            $class::created(function ($model) {
+                if (! empty($model->unsavedMedias)) {
```
Comment (freekmurze) [inline src/FileAdder/FileAdder.php]: Refactor this to bail out with a `return` statement if `model->unsavedMedias` is empty. The the actual process loses an indentation level.

File: src/FileAdder/FileAdder.php [hunk 6]
```diff
@@ -331,4 +324,30 @@ protected function sanitizeFileName(string $fileName) : string
...
+                    foreach ($model->unsavedMedias as $unsavedMedia) {
+                        $model->media()->save($unsavedMedia);
+                        $this->filesystem->add($this->pathToFile, $unsavedMedia, $this->fileName);
+                    }
+                }
+            });
+        } else {
+            $this->subject->media()->save($media);
```
Comment (freekmurze) [inline src/FileAdder/FileAdder.php]: These two lines appear twice in this function. Let's extract them to their own function. No sure on a good name here.

**Comment by ModernArtisan** [comment]
@freekmurze Can you please clarify what exactly do you mean? 

> delete all stray media items that are not related to a model

Because if the save method was never called, media model record hasn't be stored into database.

**Comment by freekmurze** [comment]
My train of thought was wrong. You're right, there's just nothing to clean. 👍

File: src/HasMedia/HasMediaTrait.php [hunk 7]
```diff
@@ -411,7 +413,9 @@ protected function mediaIsPreloaded(): bool
      */
     public function loadMedia(string $collectionName)
     {
-        return $this->media
+        $collection = $this->exists ? $this->media : collect($this->unsavedMediaItems);
```
Comment (freekmurze) [inline src/HasMedia/HasMediaTrait.php]: Could you add a testcase that proves this works when the model hasn't been saved yet?

**Comment by ModernArtisan** [comment]
@freekmurze Testcase added, something else?

**Comment by freekmurze** [comment]
Thank you very much for this! 👍

**Pull request merged**